### PR TITLE
bugfix : Incorrect sample division for "Single-Node, Multi-GPU" ddp training

### PR DESCRIPTION
**Pull request by robinnarsinghranabhat**
## Issue
https://github.com/stanford-futuredata/ColBERT/issues/403

## Main Fix 
Training Samples i.e `<query, positive_doc, negative_doc>` triples, is not correctly divided during "single-node, multi-gpu" training.

## What I checked to verify correctness

The [colbert v2 training snippet](https://github.com/stanford-futuredata/ColBERT?tab=readme-ov-file#advanced-training-colbertv2-style) from readme was used. With batch-size of 32, I ran a few iterations of training in a `single-node, two-gpu` setup. 

I added breakpoint at line 104 of `colbert/training/training.py`, and checked the scores for the first batch in two different training runs, setting `nranks` at 1 and 2. 

```
colbert.eval() ## In train mode, Dropout layers will effect output 
scores = colbert(*encoding)
```

## Training Script 
```
from colbert.infra import Run, RunConfig, ColBERTConfig
from colbert import Trainer

if __name__=='__main__':
    with Run().context(RunConfig(nranks=1, experiment="msmarco1gpu")):

        config = ColBERTConfig(
            bsize=32,
            root="/home/usd.local/robin.ranabhat/retriever/ColBERT/experiments/",
        )
        trainer = Trainer(
            triples="colbertv2.0_msmarco_64way/examples.json", # 28G, downloaded from huggingface
            # https://microsoft.github.io/msmarco/Datasets#passage-ranking-dataset
            queries="queries.train.tsv", # queries.tar.gz
            collection="collection.tsv", # collection.tar.gz
            config=config,
        )

        checkpoint_path = trainer.train()

        print(f"Saved checkpoint to {checkpoint_path}...")
```

Scores were nearly similar as expected. Minor differences should be due to default `config.amp=True`. 

## GPU Details
```
(colbert) [robin.ranabhat@usd.local@gpu001 ColBERT]$ nvidia-smi
Wed Sep 10 22:40:44 2025
+---------------------------------------------------------------------------------------+                                           
| NVIDIA-SMI 545.23.08              Driver Version: 545.23.08    CUDA Version: 12.3     |                                           
|-----------------------------------------+----------------------+----------------------+                                           
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |                                           
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |                                           
|                                         |                      |               MIG M. |                                           
|=========================================+======================+======================|                                           
|   0  Tesla P100-PCIE-16GB           Off | 00000000:18:00.0 Off |                    0 |                                           
| N/A   37C    P0              32W / 250W |      0MiB / 16384MiB |      0%      Default |                                           
|                                         |                      |                  N/A |                                           
+-----------------------------------------+----------------------+----------------------+                                           
|   1  Tesla P100-PCIE-16GB           Off | 00000000:AF:00.0 Off |                    0 |                                           
| N/A   36C    P0              29W / 250W |      0MiB / 16384MiB |      1%      Default |                                           
|                                         |                      |                  N/A |                                           
+-----------------------------------------+----------------------+----------------------+                                           
                                                                                                                                    
+---------------------------------------------------------------------------------------+                                           
| Processes:                                                                            |                                           
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |                                           
|        ID   ID                                                             Usage      |                                           
|=======================================================================================|                                           
|  No running processes found                                                           |                                           
+---------------------------------------------------------------------------------------+    
```

### CODE REVIEW

**Comment by vishalbakshi** [comment]
Thanks @robinnarsinghranabhat. As an additional check I wanted to recreate this training on 4+ GPUs. Just confirming that [this is the dataset you used](https://huggingface.co/colbert-ir/colbertv2.0_msmarco_64way/tree/main)?

**Comment by robinnarsinghranabhat** [comment]
Ahh yes :) 
And other files in that script from :
https://microsoft.github.io/msmarco/Datasets#passage-ranking-dataset

**Comment by robinnarsinghranabhat** [comment]
@vishalbakshi  any chances you got time to run this test :)

**Comment by vishalbakshi** [comment]
> @vishalbakshi any chances you got time to run this test :)

Not yet. My conservative estimate is that I'll get to this in October. Working on dependency management right now (PyTorch 2.x) for the next release. I will try to review and merge this PR before the next release though.

**Comment by vishalbakshi** [comment]
@robinnarsinghranabhat Thanks for your patience! I ran training runs with your bugfix on 4 and 8-GPU instances and I think it looks good.

## 4-GPU

With your bugfix---loss decreases, pos score increase, neg score decreases; all good things.

original_run1_vs_bugfix-4GPUs

## 8-GPU

Same trend---bugfix improves all metrics. <mark>The 8-GPU loss curve starts lower than the 4-GPU loss curve, but I don't think that's necessarily a bad thing?</mark>.

original_run1_vs_bugfix-8GPUs

## Sanity Check: Running Original Runs Twice

As a sanity check, I ran each "original" (i.e. without bugfix) training run twice just to make sure that there aren't isn't any randomness introduced causing the decrease in loss curves. <mark>Runs are identical so I think we can trust that the bugfix improvements are real</mark>.

### 4-GPU

original_run1_vs_run2-4GPUs

### 8-GPU

image


---

<mark>If this all looks good to you, I'll merge this PR. Otherwise, let me know if we should run additional tests. Thanks!</mark>

**Comment by robinnarsinghranabhat** [comment]
Hi @vishalbakshi  I am using this updated bugfix for my experiments as well. it's good to merge :)

Appreciate this extensive testing man !

In near future, I want to work on a multi-node multi-gpu support. What do you think on that ? 
Also, if there's some bugfix I can work on, Iet me know. happy to help

**Comment by vishalbakshi** [comment]
@robinnarsinghranabhat  Excellent, thanks for the confirmation and the PR! I will merge shortly.

I have been hyper-focused on the PyTorch 2.x upgrade so I haven't taken time to look ahead to bug fixes beyond my initial review, so I have yet to prioritize them. That being said, the following issues are multi-gpu related if you want to explore them:

- https://github.com/stanford-futuredata/ColBERT/issues/132
- https://github.com/stanford-futuredata/ColBERT/issues/158
- https://github.com/stanford-futuredata/ColBERT/issues/213
- https://github.com/stanford-futuredata/ColBERT/issues/265
- https://github.com/stanford-futuredata/ColBERT/issues/318
- https://github.com/stanford-futuredata/ColBERT/issues/331
- https://github.com/stanford-futuredata/ColBERT/issues/348
- https://github.com/stanford-futuredata/ColBERT/issues/349
- https://github.com/stanford-futuredata/ColBERT/issues/387

My next priority on the Roadmap after PyTorch 2.x is to replace faiss with fastkmeans, and that will be probably keep me busy deep into November. 

Multi-node multi-GPU support is definitely useful, but I would place that under the Roadmap's [Long-Term Goals: Feature Requests](https://github.com/stanford-futuredata/ColBERT/blob/main/ROADMAP.md#long-term-goals-feature-requests-3-months) and it's TBD which feature requests we'll prioritize. Please feel free to work on it, I just want to set expectations. Thanks!
